### DISCUSSION TITLE
Multiple Gitlab pipelines per commit

### DISCUSSION THREAD

**Discussion by bobvandevijver** [renovatebot/renovate#8759] [Request Help]
So, I'm basically experiencing what is described in #5743: I get two pipelines for every commit in Gitlab when using the `stabilityDays` functionality.

![image](https://user-images.githubusercontent.com/1835343/108358927-9a03e100-71ef-11eb-9c62-fd6e705fa4d2.png)

This actually already caused an MR to be merged while the pipeline wasn't completed yet. Note that this happened after an automatic rebase to make the change directly mergable.

![image](https://user-images.githubusercontent.com/1835343/108358991-b1db6500-71ef-11eb-902e-321ea5ad7153.png)

As, as you can see, the renovate pipeline was considered to be enough.

![image](https://user-images.githubusercontent.com/1835343/108359218-04b51c80-71f0-11eb-8ca5-ba396e551968.png)

This means that even when my own pipeline fails, renovate will still be able to merge due to the simple stability check being green.

When looking at the [API documentation of Gitlab](https://docs.gitlab.com/ee/api/commits.html#post-the-build-status-to-a-commit) and testing it myself on a test branch with MR and pipeline, it seems to work as you would expect: the status is added to the existing pipeline for the commit sha.

That leads me to think that this is just a simple race condition, and that just waiting a bit for the pipeline creation would already solve this. Any opinions?

**Comment by bobvandevijver** [comment]
A possible solution would be to not use the commit status at all: the check is already done in the renovate runner, so if it is converted into a checkbox in the MR description (which can be positioned under the "automerge" imo) it would be very nice.

That way, a developer can easily choose to skip the `stabilityDays` check by checking the box manually, and otherwise we wait for the next renovate run that checks the box when the stability condition is met (so renovate can never untick the box after having created it).

The state of the checkbox can then be used to determine whether or not the MR is automerge ready.

I do believe this might be an excellent feature for every platform.

**Reply by rarkins** [reply]
Stability days isn't used only for automerging though. It's set as a status check to discourage users from merging either, but they may review.

**Reply by bobvandevijver** [reply]
Sure, but currently that doesn't work as intended in Gitlab as it doesn't have the real status checks as Github does have. Most certainly not when you require the pipeline to complete successfully before merge.

That's why I would expect to see some status message in the description (or maybe as comment) from the bot about the stability days: your own pipeline is in that case the only one that runs and can be used for the Gitlab requirement.

That also means that renovate shouldn't try automerging until the stability days has been met (it does check that during its run right?). It can directly update the description or post a comment stating that the stability days have been met. That way, we won't even need the checkbox.

And this can well be Gitlab only behaviour imho.

**Reply by viceice** [reply]
maybe a small delay like in `bitbucket-server` would solve this issue until gitlab will solve this?

https://github.com/renovatebot/renovate/blob/3e2d9ca0e7596cf978acee8133a7b3b9ca1020e8/lib/platform/bitbucket-server/index.ts#L372-L377

**Reply by bobvandevijver** [reply]
I believe that might indeed be a solution to solve the double pipeline creation

**Reply by bobvandevijver** [reply]
The screenshot below illustrates the race condition quite well I guess: this particular MR has been rebased twice, but only once the status check created a separate pipeline.

![image](https://user-images.githubusercontent.com/1835343/108528195-7023ea80-72d3-11eb-987f-8d2cfa6eb907.png)

I will create a small PR that adds a small delay before the Gitlab API is called to add the branch status.

**Reply by bobvandevijver** [reply]
With the change from #8771 the placement of the stability days check is much, much more predicable 😃  
It is not 100%, but that is probably due to the fact that the onboarding process is still running, so a lot of branches, prs and pipelines are created at the same time.
